lib/MQTT/MqttClient.cpp

```diff
@@ -0,0 +1,198 @@
+    buffer[len] = '\0';
+  }
+};
+
+MqttClient::MqttClient(const Settings &settings, WiFiClient &client)
+    : settings(settings), mqttClient(new PubSubClient(client)) {
+  DebugLn(F("New MQTT client instance."));
+}
+
+MqttClient::~MqttClient() {
+  if (mqttClient) {
+    mqttClient->disconnect();
+    delete mqttClient;
+  }
+  const SetHandler *current = setHandlers;
+
+  while (current != nullptr) {
+    const SetHandler *tmp = current;
+    current = current->next;
+    delete (tmp);
+  }
+}
+
+void MqttClient::begin() {
+  using namespace std::placeholders;
+  mqttClient->setServer(settings.mqttBroker.c_str(), settings.mqttBrokerPort);
+
+  mqttClient->setCallback(std::bind(&MqttClient::onMessage, this, _1, _2, _3));
+
+  reconnect();
+}
+
+bool MqttClient::connect() {
+  if (0 == settings.mqttUser.length()) {
+    return mqttClient->connect(settings.deviceName.c_str(),
+                               settings.deviceName.c_str(), 0, true, "offline");
+  } else {
+    return mqttClient->connect(settings.deviceName.c_str(),
+                               settings.mqttUser.c_str(),
+                               settings.mqttPassword.c_str(),
+                               settings.deviceName.c_str(), 0, true, "offline");
+  }
+}
+
+bool MqttClient::reconnect() {
+  if (!mqttClient->connected()) {
+    if (connect()) {
+      DebugLn(F("MQTT connected"));
+      if (subsrcibe()) {
+        mqttClient->publish(settings.deviceName.c_str(), "online", true);
+        DebugLn(F("Subscribed!"));
+      } else {
+        DebugLn(F("Cannot subsrcibe!"));
+      }
+    } else {
+      DebugLn(F("Cannot connect!"));
+      return false;
+    }
+  }
+  return true;
+}
+
+bool MqttClient::subsrcibe() {
+  DebugLn(F("...Subscribe to: "));
+  DebugLn((settings.mqttOtaTopic + "+").c_str());
+  DebugLn((settings.mqttConfigTopic + "+").c_str());
+  DebugLn((settings.mqttSendTopic + "+").c_str());
+
+  return mqttClient->subscribe((settings.mqttOtaTopic + "+").c_str()) &&
+         mqttClient->subscribe((settings.mqttConfigTopic + "+").c_str()) &&
+         mqttClient->subscribe((settings.mqttSendTopic + "+").c_str());
+}
+
+void MqttClient::loop() {
+  if (reconnect()) {
+    mqttClient->loop();
+  }
+}
+
+void MqttClient::onMessage(char *topic, uint8_t *payload, unsigned int length) {
+  PayloadString strPayload(payload, length);
+  String strTopic(topic);
+
+  Debug(F("New MQTT message: "));
+  Debug(strTopic);
+  Debug(F(" .. "));
+  DebugLn(strPayload);
+
+  if (strTopic.startsWith(settings.mqttSendTopic)) {
+    if (codeCallback) {
+      codeCallback(String(topic + settings.mqttSendTopic.length()), strPayload);
+    }
+  }
+
+  if (strTopic.startsWith(settings.mqttConfigTopic)) {
+    String topicPart(topic + settings.mqttConfigTopic.length());
+
+    Debug(F("Config: "));
+    DebugLn(topicPart);
+
+    const SetHandler *current = setHandlers;
+
+    while (current != nullptr) {
+      if (topicPart.startsWith(current->path)) {
```

Comment:
Why `startsWith`? IMO it should be a full match, otherwise we could get conflicts.

Reply:
Originally I planned to give the CB a part of the topic. This remained as an artifact. I can change it.